x/bep3/keeper/htlt.go

```diff
@@ -0,0 +1,248 @@
+	if err != nil {
+		return sdk.ErrInternal(err.Error())
+	}
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeDepositHtlt,
+			sdk.NewAttribute(types.AttributeKeyHtltSwapID, fmt.Sprintf("%s", swapID)),
+			sdk.NewAttribute(types.AttributeKeyCoinDenom, fmt.Sprintf("%s", coin.Denom)),
+			sdk.NewAttribute(types.AttributeKeyCoinAmount, fmt.Sprintf("%d", coin.Amount.Int64())),
+		),
+	)
+
+	// Update HTLT state
+	htlt.Amount = htlt.Amount.Add(coins)
+	currExpectedIncome, _ := sdk.ParseCoins(htlt.ExpectedIncome)
+	htlt.ExpectedIncome = currExpectedIncome.Add(coins).String()
+
+	k.SetHTLT(ctx, htlt, decodedSwapID)
+
+	return nil
+}
+
+// ClaimHTLT validates a claim attempt, and if successful, sends the escrowed amount and closes the HTLT
+func (k Keeper) ClaimHTLT(ctx sdk.Context, from sdk.AccAddress, encodedSwapID string, randomNumber []byte) sdk.Error {
+	decodedSwapID, err := types.HexEncodedStringToBytes(encodedSwapID)
+	if err != nil {
+		return sdk.ErrInternal(err.Error())
+	}
+
+	htlt, found := k.GetHTLT(ctx, decodedSwapID)
+	if !found {
+		return types.ErrHTLTNotFound(k.codespace, encodedSwapID)
+	}
+
+	// Calculate hashed random number with param number and timestamp
+	hashedRandomNumber := types.CalculateRandomHash(randomNumber, htlt.Timestamp)
+	stringRandomNumber := types.BytesToHexEncodedString(hashedRandomNumber)
+
+	// Calculate hashed secret hashed random number, htlt sender, and sender other chain
+	hashedSecret, err := types.CalculateSwapID(stringRandomNumber, htlt.From, htlt.SenderOtherChain)
+	if err != nil {
+		return sdk.ErrInternal(err.Error())
+	}
+
+	encodedHashedSecret := types.BytesToHexEncodedString(hashedSecret)
+	if encodedHashedSecret != encodedSwapID {
+		return types.ErrInvalidClaimSecret(k.codespace, encodedHashedSecret, encodedSwapID)
+	}
+
+	// If HTLT is not cross-chain, htlt.ExpectedIncome should equal htlt.Amount
+	claimerCoins, err := sdk.ParseCoins(htlt.ExpectedIncome)
+	if err != nil {
+		return sdk.ErrInternal(err.Error())
+	}
+	deputyCoins := htlt.Amount.Sub(claimerCoins)
+
+	// Send expected income from bep3 module to claiming address
+	err = k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, from, claimerCoins)
+	if err != nil {
+		return sdk.ErrInternal(err.Error())
+	}
+
+	// Send remaining amount from bep3 module to deputy
+	if htlt.CrossChain {
+		err = k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, htlt.From, deputyCoins)
+		if err != nil {
+			return sdk.ErrInternal(err.Error())
+		}
+	}
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeClaimHtlt,
+			sdk.NewAttribute(types.AttributeKeyHtltSwapID, fmt.Sprintf("%s", encodedSwapID)),
+			sdk.NewAttribute(types.AttributeKeyClaimer, fmt.Sprintf("%s", from)),
+			sdk.NewAttribute(types.AttributeKeyCoinDenom, fmt.Sprintf("%s", claimerCoins[0].Denom)),
+			sdk.NewAttribute(types.AttributeKeyCoinAmount, fmt.Sprintf("%d", claimerCoins[0].Amount.Int64())),
```

Comment:
Lines 175,176 will panic if someone submits a HTLT with expectedIncome=“” as claimerCoins will be empty.
In general I think using `sdk.Coin` when there should only be one coin is safer (but possibly more awkward to write).
Similarly `kepper.ValidateAsset` will panic if coins arg is nil. Not that it looks like this ever happens in the code, but could happen after some future refactor.

Reply:
1. Updated MsgCreateHTLT's ValidateBasic() in https://github.com/Kava-Labs/kava/pull/370/commits/fecee798f71734887c165f44e22893a4597d935a.

2. I agree that the logic here is pretty brittle. Updated ValidateAsset() in https://github.com/Kava-Labs/kava/pull/370/commits/36de1ebcaadeb0ea76f7b86709d5876f44ee1ad2. Do you have any insight into best practices for the location of these assertions? ValidateBasic() checks that the coin's amount is positive (meaning there must be at least one coin), but I feel that it may be better for the `len(coins) != 1` check to happen inside ValidateBasic() as opposed to ValidateAsset().